common/optimiser/OptimiserConstraint.java

```diff
@@ -23,36 +23,49 @@
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 
 public class OptimiserConstraint {
 
+    private final Optimiser optimiser;
     private final double lowerBound;
     private final double upperBound;
     private final String name;
     private final Map<OptimiserVariable<?>, Double> coefficients;
     private MPConstraint mpConstraint;
 
-    public OptimiserConstraint(double lowerBound, double upperBound, String name) {
+    public OptimiserConstraint(Optimiser optimiser, double lowerBound, double upperBound, String name) {
+        this.optimiser = optimiser;
         this.lowerBound = lowerBound;
         this.upperBound = upperBound;
         this.coefficients = new HashMap<>();
         this.name = name;
     }
 
     public void setCoefficient(OptimiserVariable<?> variable, double coeff) {
-        coefficients.put(variable, coeff);
+        assert mpConstraint == null || coefficients.containsKey(variable);
```

Comment:
Why would adding a coefficient in a changed constraint be an error we want to assert against?

Reply:
i'm more strictly defining the type of solving we support as "The coefficients may be dynamic, but not the number of terms in constraints or objective"